### PR TITLE
Fix RoutingInfo consistency

### PR DESCRIPTION
**Pull request by wprzytula**
A bug has been found that put wrong `Consistency` and `SerialConsistency` into `RoutingInfo`. In order to:
- be sure that it's fully fixed,
- prevent regressions,
- check consistencies in other contexts,

the following is done:
- proxy acquires deserialization capabilities for request frames; it understands `Query`, `Execute` and `Batch` frames,
- a test is added to check that request frames contains consistencies that match the specified ones,
- a test is added to check that `RoutingInfo` for `LoadBalancingPolicy` contains the specified consistencies,
- as the latter test confirms the bug in `Session`, the affected methods are fixed,
- minor refactor is done in `Session` to make fiddling with consistency clearer,
- bonus: utilities for single-node dry-mode cluster are added, based on the bindings' test implementation. These will most probably prove useful for writing future tests.

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- ~~[ ] I have adjusted the documentation in `./docs/source/`.~~
- ~~[ ] I added appropriate `Fixes:` annotations to PR description.~~

### CODE REVIEW

**Review by piodul — CHANGES_REQUESTED** [review]
While the purpose of the PR is good, I think it is too bloated. There is a lot of code duplication in tests and proxy utilities, not to mention the last commit which doesn't really belong to this PR.

- Request deserialization code should be put into scylla-cql, as it belongs there and not in scylla-proxy. At least, I don't think that we need to duplicate the structure definitions.
- There is lots of code to be deduplicated or shortened in the tests that this PR adds,
- The last commit needs to go out and be resent separately. I don't see how it belongs here.

Please try to reduce the number of lines added by this PR. I don't believe we need a nearly 1000-line PR to fix an issue about some parameters not being propagated properly in a handful of functions.

File: scylla/tests/integration/consistency.rs [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+    [SerialConsistency::Serial, SerialConsistency::LocalSerial]
+}
+
+// This is just to increase the likelihood that only intended prepared statements (which contain this mark) are captured by the proxy.
+const MAGIC_MARK: i32 = 123;
+
+#[tokio::test]
+#[ntest::timeout(60000)]
+#[cfg(not(scylla_cloud_tests))]
+async fn consistency_is_correctly_set_in_cql_requests() {
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla/tests/integration/consistency.rs:234]: Is it readable enough now?

File: scylla/tests/integration/utils.rs [hunk 2]
```diff
@@ -116,3 +119,55 @@ where
...
+#[allow(unused)]
+// As these are very generic, they should be put last in the rules Vec.
+pub(crate) fn generic_drop_queries_rules() -> impl IntoIterator<Item = RequestRule> {
+    [RequestRule(
+        Condition::RequestOpcode(RequestOpcode::Query),
+        // We won't respond to any queries (including metadata fetch),
+        // but the driver will manage to continue with dummy metadata.
+        RequestReaction::drop_connection(),
+    )]
+}
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla/tests/integration/utils.rs]: For now, I've removed the commit from this PR.

File: scylla/tests/integration/consistency.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla/tests/integration/consistency.rs:236]: ping @piodul

[the remark this answers is not in the document]
Comment (piodul) [reply scylla/tests/integration/consistency.rs:236]: I'm not sure. The tests in this file are quite long and the flow is hard to follow, but I don't have good suggestions on how to improve it.

File: scylla/src/transport/session.rs [hunk 3]
```diff
@@ -1593,7 +1593,7 @@ impl Session {
                             &execution_profile,
                             ExecuteQueryContext {
                                 is_idempotent: statement_config.is_idempotent,
-                                consistency: statement_config.consistency,
+                                consistency_set_on_statement: statement_config.consistency,
```
Comment (piodul) [inline scylla/src/transport/session.rs:1596]: > As there happens a lot with consistencies in session's methods that
> execute requests, proper naming is crucial. Therefore,
> ExecuteQueryContext's field is renamed from the vague `consistency` to
> `consistency_set_on_statement` (the one directly set), as it is exactly
> its semantics.

I don't understand the first sentence and the implication that follows it.

File: scylla-cql/src/frame/request/batch.rs [hunk 4]
```diff
@@ -14,6 +14,7 @@ use crate::frame::{
 const FLAG_WITH_SERIAL_CONSISTENCY: u8 = 0x10;
 const FLAG_WITH_DEFAULT_TIMESTAMP: u8 = 0x20;
 
+#[derive(Debug)]
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla-cql/src/frame/request/batch.rs:18]: I've seen many times in discussions over Rust codebases that `Debug` should be always available whenever possible (and we are not in a strongly limited environment such as embedded). For instance: https://github.com/rust-lang/rust/pull/32054#issuecomment-192564644

File: scylla/tests/integration/consistency.rs [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+                        Condition::or(
+                            Condition::RequestOpcode(RequestOpcode::Batch),
+                            Condition::and(
+                                Condition::RequestOpcode(RequestOpcode::Query),
+                                Condition::BodyContainsCaseSensitive(Box::new(
+                                    *b"INTO consistency_tests",
+                                )),
+                            ),
+                        ),
+                    ),
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/tests/integration/consistency.rs:253]: https://doc.rust-lang.org/std/ops/index.html

> Note that the && and || operators are currently not supported for overloading. Due to their short circuiting nature, they require a different design from traits for other operators like [BitAnd](https://doc.rust-lang.org/std/ops/trait.BitAnd.html). Designs for them are under discussion.

Interesting, I didn't know that. Maybe we can consider using the `&` and `|` operators instead until they figure the logical ones out.

File: scylla-cql/src/frame/request/batch.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (piodul) [reply scylla-cql/src/frame/request/batch.rs:18]: Well, the hit in compile time is not ideal. Perhaps it makes sense to enable it for public types that users interact with directly, though I'm not sure if it we should do it for non-public ones (I treat the types from scylla-cql to be kind of non-public).

Regardless, if the addition of Debug derives is not strictly necessary for the PR then it should be sent separately.

File: scylla/tests/integration/consistency.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla/tests/integration/consistency.rs:253]: I'm in favour of explicit methods in case of combining `Condition`s. Idiomatic Rust doesn't switch to operator overloading, unlike C++.

[the remark this answers is not in the document]
Comment (piodul) [reply scylla/tests/integration/consistency.rs:253]: > I'm in favour of explicit methods in case of combining `Condition`s.

If by methods you mean non-static methods here, then I'd argue that they are similarly awkward as calling them without `self`.

> Idiomatic Rust doesn't switch to operator overloading, unlike C++.

I don't think this is true. Operators are usually overloaded for mathematical objects such as bignums or vectors. In our case, you would construct a predicate, a boolean expression - boolean operator are a natural fit (well, we would need to use the bitwise equivalents but it would be close enough). I don't think it is a big abuse of the syntax.

Ultimately, I think we should just allow closure predicates that takes a context object which allows doing all those checks directly in the code (e.g. `|ctx| match ctx.request_opcode() { ... }`). This would lead to much more natural syntax and greater flexibility.

File: scylla-cql/src/frame/request/batch.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla-cql/src/frame/request/batch.rs:18]: Regarding visibility of `scylla-cql` entities, I've always had an approach that this is a low-level crate that could be used by developers on its own for their own purposes, and implementing `Debug` may come in handy for them.

File: scylla-cql/src/frame/request/batch.rs [hunk 6]
```diff
@@ -1,20 +1,24 @@
...
-    value::BatchValues,
+    value::{BatchValues, SerializedValues},
 };
 
 // Batch flags
 const FLAG_WITH_SERIAL_CONSISTENCY: u8 = 0x10;
 const FLAG_WITH_DEFAULT_TIMESTAMP: u8 = 0x20;
 
-pub struct Batch<'a, StatementsIter, Values>
+pub struct BatchOwnershipGeneric<Query, Prepared, StatementsIter, Values>
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla-cql/src/frame/request/batch.rs:19]: The only reason why I didn't adopt the naming you suggest here is that I wanted to save the existing naming for the borrowed version.

File: scylla-cql/src/frame/request/batch.rs [hunk 7]
```diff
@@ -42,7 +42,7 @@ pub type BatchOwned = BatchOwnershipGeneric<
 >;
 
 /// The type of a batch.
-#[derive(Clone, Copy, Debug)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla-cql/src/frame/request/batch.rs:46]: This is required for implementing `PartialEq` for `BatchOwned`, which is used in tests.

File: scylla-cql/src/frame/request/mod.rs [hunk 8]
```diff
@@ -48,3 +48,137 @@ pub trait Request {
...
+                .zip(other.statements.clone())
+                .all(|(fst, snd)| fst == snd)
+                && self.statements_count == other.statements_count
+                && self.batch_type == other.batch_type
+                && self.consistency == other.consistency
+                && self.serial_consistency == other.serial_consistency
+                && self.timestamp == other.timestamp
+                && self.values == other.values
+        }
+    }
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla-cql/src/frame/request/mod.rs:104]: As far as I understand, `#[derive()]` does not support deriving implementations only for some specialisations of generic type (in this case, specialisations whose `Values` type parameter implements `PartialEq`).

File: scylla/src/transport/session.rs (hunk 3 above)
Comment (wprzytula) [reply scylla/src/transport/session.rs:1596]: Let me rephrase it this way:

> There are three levels of consistency that are taken into account by the `Session`: default exec profile's, per-statement exec profile's and per-statement one. To avoid confusion, we should use naming that clearly distinguishes them. Therefore, ...

Is it clear now?

File: scylla-cql/src/frame/value.rs [hunk 9]
```diff
@@ -197,6 +198,40 @@ impl SerializedValues {
...
+        let mut buf = &self.serialized_values[..];
+        (0..self.values_num).map(move |_| {
+            // `unwrap()`s here are safe, as we assume type-safety: if `SerializedValues` exits,
+            // we have a guarantee that the layout of the serialized values is valid.
+            let name = self
+                .contains_names
+                .then(|| types::read_string(&mut buf).unwrap());
+            let serialized = types::read_bytes(&mut buf).unwrap();
+            (name, serialized)
+        })
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla-cql/src/frame/value.rs:233]: I don't understand how I can refactor this to be more readable. `for` is not applicable as I want to return an iterator, and `then` is an elegant way to create an `Option` which I need to return eventually.

File: scylla/src/transport/session.rs (hunk 3 above)
Comment (piodul) [reply scylla/src/transport/session.rs:1596]: Yes, sounds better.

File: scylla-cql/src/frame/request/batch.rs [hunk 10]
```diff
@@ -34,6 +34,22 @@ pub enum BatchType {
     Counter = 2,
 }
 
+impl TryFrom<u8> for BatchType {
+    type Error = ParseError;
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla-cql/src/frame/request/batch.rs:54]: Is this really resolved?

[the remark this answers is not in the document]
Comment (piodul) [reply scylla-cql/src/frame/request/batch.rs:54]: Ok, you probably didn't push a new version, nvm

File: scylla-cql/src/frame/request/batch.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (piodul) [reply scylla-cql/src/frame/request/batch.rs:46]: Ok, then please include this in the description of the commit, not here in a GitHub comment.

File: scylla-cql/src/frame/value.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (piodul) [reply scylla-cql/src/frame/value.rs:233]: Ok, I didn't notice that this function returns an iterator. Maybe if Rust had generators stabilised then a transformation similar to the one in `new_from_frame` would be feasible. In that case, it's OK.

File: scylla-cql/src/frame/request/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (piodul) [reply scylla-cql/src/frame/request/mod.rs:104]: Did you try it? Derive macros operate on the whole definition of a struct, so in theory it could take the trait bound into account.

If not, then maybe we could drop this bound?

Alternatively, if we use `Cow` then the problem should solve itself.

File: scylla-cql/src/frame/request/batch.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (piodul) [reply scylla-cql/src/frame/request/batch.rs:18]: Our main priority should be the `scylla` crate. We don't give as much attention or care about stability of the `scylla-cql`'s API. The main reason we created it in the first place was because we needed it for `scylla-proxy`.

File: scylla-cql/src/frame/request/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla-cql/src/frame/request/mod.rs:104]: Derive macro means: "implement the trait automatically. Dot.". It does not say: "implement the trait for all generic parameters that implement the trait as well", and does not allow for this.

[the remark this answers is not in the document]
Comment (piodul) [reply scylla-cql/src/frame/request/mod.rs:104]: A derive macro has access to the whole definition of a struct, including the generic parameters and their trait bounds. Such a macro can copy those bounds into the generated `impl` block. For example, the derive macro I implemented [here](https://github.com/scylladb/scylla-rust-driver/pull/665/commits/c6e08762a8476b4e938b16152fb65941a21a5ae0) takes the trait bounds of each parameter into account. Therefore, it is allowed. The question is just whether the stock derive macros do this.

File: scylla-cql/src/frame/request/batch.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla-cql/src/frame/request/batch.rs:18]: I'll make the derive `cfg_attr(test)`, as it's needed for tests.

File: scylla/tests/integration/consistency.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla/tests/integration/consistency.rs:253]: This is anyway out of this PR's scope. For now, let's stick to methods.

File: scylla-cql/src/frame/request/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (piodul) [reply scylla-cql/src/frame/request/mod.rs:147]: Can we change `statements` not to be an iterator? Can it be `Cow<'x, [T]>` where `T: SomeTraitThatConvertsToBatchFromScyllaCql`?

[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla-cql/src/frame/request/mod.rs]: Yes! I don't believe it after the lifetime storm I had to overcome, but I managed to accomplish that.

File: scylla-cql/src/frame/request/batch.rs [hunk 11]
```diff
@@ -12,12 +12,13 @@ use crate::frame::{
...
+pub struct Batch<'b, Statement, Values>
 where
-    StatementsIter: Iterator<Item = BatchStatement<'a>> + Clone,
+    BatchStatement<'b>: From<&'b Statement>,
+    [Statement]: ToOwned,
     Values: BatchValues,
 {
-    pub statements: StatementsIter,
+    pub statements: Cow<'b, [Statement]>,
     pub statements_count: usize,
```
Comment (piodul) [inline scylla-cql/src/frame/request/batch.rs:26]: Do we need `statements_count` anymore?

File: scylla-cql/src/frame/request/mod.rs [hunk 12]
```diff
@@ -48,3 +48,195 @@ pub trait Request {
...
+            parameters,
+        };
+
+        {
+            let mut buf = Vec::new();
+            query.serialize(&mut buf).unwrap();
+
+            let query_deserialized = Query::deserialize(&mut &buf[..]).unwrap();
+            assert_eq!(&query_deserialized.contents, &query.contents);
+            assert_eq!(&query_deserialized.parameters, &query.parameters);
```
Comment (piodul) [inline scylla-cql/src/frame/request/mod.rs:148]: Why not `assert_eq!(query_deserialized, query);`?

Comment (wprzytula) [reply scylla-cql/src/frame/request/mod.rs:148]: This would require `PartialEq`.

File: scylla/src/transport/session.rs [hunk 13]
```diff
@@ -1147,19 +1173,37 @@ impl Session {
...
                     keyspace: ps.get_keyspace_name(),
                     is_confirmed_lwt: false,
                 }
             }
-            _ => RoutingInfo::default(),
+            _ => RoutingInfo {
+                consistency,
+                serial_consistency,
+                ..Default::default()
+            },
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla/src/transport/session.rs:1208]: Indeed, it does. I've fixed it and extended the routing info test to cover `query_iter`/`execute_iter` cases too.

File: scylla-cql/src/frame/request/mod.rs (hunk 12 above)
Comment (wprzytula) [reply scylla-cql/src/frame/request/mod.rs:148]: I've derived `PartialEq` under `test` cfg attr.

**Review by piodul — APPROVED** [review]
